---
title: "1_load_normalize"
author: "shaphan"
date: "11/17/2022"
pdf_document: default
html_document: default
     toc: true
     keep_md: true
editor_options: null
chunk_output_type: console
---


# 0 - load libraries
## Remove contaminant sequences data to improve taxonomic representation in samples 
```{r}

#load necessary libraries
library(phyloseq)
library(ggplot2)
library(vegan)
library(tidyr)
library(decontam)
library("metagenomeSeq")


# increases memory limit used by R
memory.limit(size = 350000)

# custom functions related to data loading and decontamination
source("./Code/functions_loading_and_decontamination.R")

```

# 1 - prepare 16s data

## 1.1 - load 16S data and build Bac phyloseq object
```{r}


# loads the OTU table, immediately saving it as a phyloseq OTU table (lighter than a data frame)

raw_bac_otutab<-otu_table(object = read.table(file = "./Data/16S_ECA_seqtab_final.txt",
                                              header = TRUE,
                                              row.names = 1, # first column has row names (ASV names)
                                              check.names = FALSE), # prevents "X" to be added to column                                                                          names, such as X49_16S,
                          taxa_are_rows = TRUE)

sample_sums(raw_bac_otutab)%>%sum


#updated tax
sklearn_bac_taxtab<-read.table(file = "./Data/16S_ECA_taxonomy.tsv",
                               header = TRUE,
                               sep = "\t",
                               row.names = 1,
                               check.names = FALSE)



# adjusts number and name of columns
sklearn_bac_taxtab<- separate(data = sklearn_bac_taxtab,
                              col = Taxon,
                              into =c("Kingdom",
                                      "Phylum",
                                      "Class",
                                      "Order",
                                      "Family",
                                      "Genus",
                                      "Species"),
                              sep = "; ")


# change from d__Bacteria to k__bacteria
sklearn_bac_taxtab$Kingdom<-gsub("d__", "k__", sklearn_bac_taxtab$Kingdom)


# save taxa as phyloseq object
sklearn_bac_taxtab<-tax_table(object = as.matrix(sklearn_bac_taxtab))


#change name and remove old object
raw_bac_taxtab<-sklearn_bac_taxtab
rm(sklearn_bac_taxtab)


# loads the representative sequences table, immediately saving it as a phyloseq OTU table (lighter than a data frame)

raw_bac_refseq<-refseq(physeq = Biostrings::readDNAStringSet(filepath = "./Data/16S_ECA_repset.fasta", use.names = TRUE))
taxa_names(raw_bac_refseq)<-gsub(" .*", "", taxa_names(raw_bac_refseq)) # drops taxonomy from ASV names


# loads the mapping tile, immediately saving it as a phyloseq OTU table (lighter than a df)
raw_bac_metadata<-sample_data(object = read.table(file = "./Data/16S_ECA_Mapping_file.txt",
                                                  header = TRUE,
                                                  sep = "\t",
                                                  row.names = 1,
                                                  check.names = FALSE))

```

## 1.2 - adjust 16S sample names

The sample names on the OTU table do not match the sample names on the metadata. this will remove unnecessary characters from the sample names on the otu tables

```{r}

# remove any number of numeric characters([0-9]+) followed by the constant (_ID2577_16S-) and then any number of numeric characters up to and including a "-" ([0-9]+-)

otutab_sampnames<-gsub("[0-9]+_ID2577_16S-[0-9]+-", "", phyloseq::sample_names(raw_bac_otutab))


#now remove anything (*) after and including the dash (-) the dot indicates that * has a special meaning (matches any character)

otutab_sampnames<-gsub("-.*", "", otutab_sampnames)


#overwrite the old sample names with the correct (short) version that was just created

phyloseq::sample_names(raw_bac_otutab)<-otutab_sampnames


#check dimensions
dim(raw_bac_otutab)
dim(raw_bac_metadata)

```

## 1.3 - build bac phyloseq object

```{r}
# build the main 16S phyloseq object by putting all these phyloseq-class objects (otu table, tax table, ref seq, metadta) into a single ps

raw_bac_ps<-merge_phyloseq(raw_bac_otutab,
                           raw_bac_taxtab,
                           raw_bac_refseq,
                           raw_bac_metadata)

#Are there missing samples?
sample_names(raw_bac_otutab)[which(!sample_names(raw_bac_otutab) %in%  sample_names(raw_bac_ps) )]


# remove old objects to reduce memory use
rm(raw_bac_otutab,raw_bac_taxtab,raw_bac_refseq,raw_bac_metadata)


# change names from ASV to  bASV, so they can be distinguished from bacterial ASVs

taxa_names(raw_bac_ps)<-paste("b", taxa_names(raw_bac_ps), sep = "")
tax_table(raw_bac_ps)<-gsub(" ", "", tax_table(raw_bac_ps)) # drops a space character from taxa names. I don't know how that character ended up in there


# let's check the imported objects. Often errors will arise from typos when filling up the data sheets

otu_table(raw_bac_ps)[1:10,1:10]
sample_data(raw_bac_ps)[1:10,1:10]
tax_table(raw_bac_ps)[1:10,1:6]
refseq(raw_bac_ps)

# run garbage collection after creating large objects
gc()

```

## 1.4 - remove bad Bac taxa

```{r}

# here we see that sequences of only 50 base pairs (bp) were still retained by the updated dada2 pipeline
hist(as.data.frame(refseq(raw_bac_ps)@ranges)$width, breaks = 300)
summary(as.data.frame(refseq(raw_bac_ps)@ranges)$width) # ps = phyloseq


# here we get rid of sequences shorter than 380 bp. this reduced the total number of taxa by ~ 4k (from 384089 to 380244 taxa)
raw_bac_ps<-prune_taxa(taxa = as.data.frame(refseq(raw_bac_ps)@ranges)$width>380,
                       x = raw_bac_ps)


# keep only ASVs identified as bacterial
raw_bac_ps<-subset_taxa(raw_bac_ps, Kingdom=="k__Bacteria") # drops from 380244 to 380019 taxa, barely any difference


# define library sizes as metadata before filtering
raw_bac_ps@sam_data$library_sizes_prefiltering<-sample_sums(raw_bac_ps)


# remove taxa occurring less than 21 times in the data set, as VSEARCH standard
otu_table(raw_bac_ps) <- otu_table(raw_bac_ps)[which (rowSums(otu_table(raw_bac_ps)) > 21),] # drops from 380K to 168K taxa, from 222M to 221M sequences (843k sequences removed)


# filter ASV according samples (min 0 reads, present in 7 samples)
filter <- phyloseq::genefilter_sample(raw_bac_ps, filterfun_sample(function(x) x > 0), A = 7)
raw_bac_ps <- prune_taxa(filter, raw_bac_ps) #drops from 168k to 72k taxa, drops from 221M  to 219M reads


#define plastid, mitochondria and host plant contamination ps objects
Mitochondria_ps<-subset_taxa(raw_bac_ps, Family == "f__Mitochondria" | Family == "Mitochondria")
Plastid_ps<-subset_taxa(raw_bac_ps, Order == "o__Chloroplast" | Order == "Chloroplast")
host_plant_ps<-merge_phyloseq(Mitochondria_ps,Plastid_ps)


#quick histogram showing plant DNA contamination
hist(sample_sums(host_plant_ps)/sample_sums(raw_bac_ps)*100, breaks = 100)

# define host plant 16S contamination as metadata
raw_bac_ps@sam_data$Mitochondria_reads<-sample_sums(Mitochondria_ps)
raw_bac_ps@sam_data$Plastid_reads<-sample_sums(Plastid_ps)
raw_bac_ps@sam_data$Host_DNA_n_reads<-sample_sums(host_plant_ps)
raw_bac_ps@sam_data$Host_DNA_contamination_pct<-sample_sums(host_plant_ps)/sample_sums(raw_bac_ps)*100


# remove host plant sequences (plastid and mitochondrial DNA)
raw_bac_ps<-remove_Chloroplast_Mitochondria(raw_bac_ps) # drops from 72k to 69.9k taxa, from 219M to 121M reads (almost 45% of our 16S sequencing effort! that's about 12k Euro!)


# Rename ps object
unnormalized_bac_ps<-raw_bac_ps
raw_bac_ps<-unnormalized_bac_ps


# run garbage collection after creating large objects
gc()

```

# 2 - prepare ITS data

## 2.1 - load ITS data and build Fungi (Fun) phyloseq object

```{r}

# loads the OTU table, immediately saving it as a phyloseq OTU table (lighter than a df)
raw_fun_otutab<-otu_table(object = read.table(file = "./Data/ITS_ECA_seqtab_final.txt", 
                                              header = TRUE,
                                              row.names = 1, # first column has row names (ASV names)
                                              check.names = FALSE), # prevents "X" to be added to column names, such as X49_ITS,
                           taxa_are_rows = TRUE)


#updated tax
sklearn_fun_taxtab<-read.table(file = "./Data/ITS_ECA_taxonomy.tsv", 
                               header = TRUE,
                               sep = "\t",
                               row.names = 1, # first column has row names (ASV names)
                               check.names = FALSE) # prevents "X" to be added to column names, such as X49_16S,

# adjusts number and name of columns
sklearn_fun_taxtab<- separate(data = sklearn_fun_taxtab,
                               col = Taxon,
                               into =c("Kingdom", 
                                       "Phylum", 
                                       "Class", 
                                       "Order", 
                                       "Family",
                                       "Genus", 
                                       "Species"),
                               sep = ";")

# saves taxa as phyloseq object
sklearn_fun_taxtab<-tax_table(object = as.matrix(sklearn_fun_taxtab))

# rename object, remove old versions
raw_fun_taxtab<-sklearn_fun_taxtab
rm(sklearn_fun_taxtab)


# load the representative sequences table, immediately saving it as a phyloseq OTU table (lighter than a df)
raw_fun_refseq<-refseq(physeq = Biostrings::readDNAStringSet(filepath = "./Data/ITS_ECA_repset.fasta", use.names = TRUE)) 
taxa_names(raw_fun_refseq)<-gsub(" .*", "", taxa_names(raw_fun_refseq)) 

# loads the mapping tile, immediately saving it as a phyloseq OTU table (lighter than a df)
raw_fun_metadata<-sample_data(object = read.table(file = "./Data/ITS_ECA_Mapping_file.txt", 
                                                  header = TRUE,
                                                  sep = "\t",
                                                  row.names = 1, 
                                                  check.names = FALSE))
```

## 2.2 - adjust ITS sample names

```{r}

#adjust sample names
# remove any number of numeric charters([0-9]+) followed by the constant (_ID2577_ITS-) and then any number of numeric characters up to and including a "-" ([0-9]+-)
otutab_sampnames<-gsub("[0-9]+_ID2577_ITS-[0-9]+-", "", phyloseq::sample_names(raw_fun_otutab)) 

#now remove anything (*) after and including the dash (-) the dot indicates that * has a special meaning (matches any character)
otutab_sampnames<-gsub("-.*", "", otutab_sampnames)

#overwrite the old sample names with the correct (short) version that was just created
phyloseq::sample_names(raw_fun_otutab)<-otutab_sampnames

```

## 2.3  - build Fun phyloseq object

```{r}
#check dimensions
dim(raw_fun_otutab)
dim(raw_fun_metadata)

#fungi has 383 samples on otu table instead of 404, is it expected? I think so because the raw sequences equal 766 (single reads) = 766/2 = 383 combined reads (samples). 

# build the main ITS phyloseq object by putting all these phyloseq-class objects (otu table, tax table, ref seq, metadata) into a single ps
raw_fun_ps<-merge_phyloseq(raw_fun_otutab,
                           raw_fun_taxtab,
                           raw_fun_refseq,
                           raw_fun_metadata)

# remove old objects to reduce memory use
rm(raw_fun_otutab,raw_fun_taxtab,raw_fun_refseq,raw_fun_metadata)

# change names from ASV to fASV, so they can be distinguished from bacterial ASVs
taxa_names(raw_fun_ps)<-paste("f", taxa_names(raw_fun_ps), sep = "")

# let's check the imported objects. Often errors will arise from typos when filling up the data sheets 
otu_table(raw_fun_ps)[1:10,1:10]
sample_data(raw_fun_ps)[1:10,1:10]
tax_table(raw_fun_ps)[1:10,1:7]
refseq(raw_fun_ps)

# run garbage collection after creating large objects
gc()

```

## 2.4 - Remove non-fungal sequences & filter ASVs
  * ~5.17% of total reads are plant ITS
  * ~5.77% of total reads are protozoa ITS
  * ~9.92% of ITS reads are unidentified

```{r}

# define library sizes as metadata before filtering
raw_fun_ps@sam_data$library_sizes_prefiltering<-sample_sums(raw_fun_ps)

# removes taxa occurring less than 21 times in the database, as VSEARCH standard
otu_table(raw_fun_ps) <- otu_table(raw_fun_ps)[which (rowSums(otu_table(raw_fun_ps)) > 21),]  #drops from 15.5K to 14.1K taxa, only 6K reads removed


# filter  samples (min 0 reads, present in 7 samples)
filter <- phyloseq::genefilter_sample(raw_fun_ps, filterfun_sample(function(x) x > 0), A = 7) 
raw_fun_ps <- prune_taxa(filter, raw_fun_ps) #drops from 14.1k to 3.1k taxa, drops from 206M  to 195M reads
rm(filter) # we no longer need this string


# removes g__Malassezia, a known human pathogen (saunders et al 2012, https://journals.plos.org/plospathogens/article?id=10.1371/journal.ppat.1002701)
raw_fun_ps<-prune_taxa(!(taxa_names(raw_fun_ps) %in% taxa_names(subset_taxa(raw_fun_ps, Genus=="g__Malassezia"))), raw_fun_ps) 

# take  all plant ITS sequences here to calculate host contamination
raw_PlantITS_ps<-subset_taxa(raw_fun_ps, Kingdom=="k__Viridiplantae") 


# keeps only ASVs identified as fungal
raw_fun_ps<-subset_taxa(raw_fun_ps, Kingdom=="k__Fungi") # drops from 18.5K to 15.5K taxa, from 206M reads to 201M reads


# define ITS host contamination as sample metadata
raw_fun_ps@sam_data$Host_DNA_contamination_pct<-sample_sums(raw_PlantITS_ps)/sample_sums(raw_fun_ps)*100
raw_fun_ps@sam_data$Host_DNA_n_reads<-sample_sums(raw_PlantITS_ps)


#check final sequence length. ITS sequences smaller than 100bp should be discarded
summary(raw_fun_ps@refseq@ranges@width)


# add library sizes as part of metadata
sample_data(raw_fun_ps)$library_size<-sample_sums(raw_fun_ps)


#remove unnecessary objects and collect garbage
rm(raw_PlantITS_ps)
gc()

```

# 3 - Decontaminate fungi and bacteria
```{r}

#samples present in bac but missing in Fun
sample_names(raw_bac_ps)[!(sample_names(raw_bac_ps)  %in% sample_names(raw_fun_ps))]


#decontaminate!
unnormalized_bac_ps<-decontaminate_and_plot(raw_bac_ps)
unnormalized_fun_ps<-decontaminate_and_plot(raw_fun_ps)


checm<-subset_samples(raw_bac_ps, sample =="MOCK")
checm@tax_table
checm@otu_table
sample_sums(checm)


# let's look into how many reads our blank had (average+sd) so we can compare those to libraries with very low number of reads
blank_reads_bac<-sort(sample_sums(subset_samples(raw_bac_ps, sample == "MOCK")))
plot(blank_reads_bac)


blank_reads_fun<-sort(sample_sums(subset_samples(raw_fun_ps, sample == "MOCK")))
plot(blank_reads_fun)


# after checking the decontamination plots and reports, we can remove them and focus on the phyloseq object
unnormalized_bac_ps<-unlist(unnormalized_bac_ps[[1]])
unnormalized_fun_ps<-unlist(unnormalized_fun_ps[[1]])








```



# 4 - make list of Bac and FUn objects
```{r}

# now let's save them as list, add names to this list, and remove unlisted ps objects to release memory
unnormalized_BacFun_ps_l<-list(unnormalized_bac_ps,unnormalized_fun_ps)
names(unnormalized_BacFun_ps_l)<-c("Bac","Fun")
rm(unnormalized_bac_ps,unnormalized_fun_ps)


#this is our list of objects
unnormalized_BacFun_ps_l

#backyp object
unnormalized_BacFun_ps_l_backup<-unnormalized_BacFun_ps_l

# let's make sure some key factors are recorded as factors, and numeric are saved as numeric
unnormalized_BacFun_ps_l<-lapply(unnormalized_BacFun_ps_l, function (x){
  x@sam_data$frass_dose<-as.factor(x@sam_data$frass_dose)
  x@sam_data$frass_treatment<-as.factor(x@sam_data$frass_treatment)
  x@sam_data$frass_source<-as.factor(x@sam_data$frass_source)
  x@sam_data$amendment<-as.factor(x@sam_data$amendment)
  x@sam_data$sample_type<-as.factor(x@sam_data$sample_type)
  
  x@sam_data$leaf_size<-as.numeric(x@sam_data$leaf_size)
  x@sam_data$shoot_biomass<-as.numeric(x@sam_data$shoot_biomass)
  
  # makes new variables
  x@sam_data$frass_treatment_and_dose<-paste(x@sam_data$frass_treatment, x@sam_data$frass_dose, sep = "_")
  x@sam_data$frass_treatment_dose_source<-paste(x@sam_data$frass_treatment, 
                                                x@sam_data$frass_dose, 
                                                x@sam_data$frass_source, 
                                                sep = "_")
  
  return(x)
  
  
})




```


# 4 - remove samples with very low library size
```{r}



# now let's check the lower end of our library sizes
sort(sample_sums(unnormalized_BacFun_ps_l$Bac))
plot(sort(sample_sums(unnormalized_BacFun_ps_l$Bac)))
hist(sort(sample_sums(unnormalized_BacFun_ps_l$Bac)))
 

sort(sample_sums(unnormalized_BacFun_ps_l$Fun))
plot(sort(sample_sums(unnormalized_BacFun_ps_l$Fun)))
hist(sort(sample_sums(unnormalized_BacFun_ps_l$Fun))) # this is not a normal distribution. ITS pre-processing has to be investigated


#remove the four Bac samples with very low library size
unnormalized_BacFun_ps_l$Bac<-subset_samples(physeq = unnormalized_BacFun_ps_l$Bac,
                      sample_sums(unnormalized_BacFun_ps_l$Bac)>=25000  )



#remove fungal samples wiht elss than 5k reads (this is before pedro ivnestigates ITS pre-processing)
unnormalized_BacFun_ps_l$Bac<-subset_samples(physeq = unnormalized_BacFun_ps_l$Bac,
                      sample_sums(unnormalized_BacFun_ps_l$Bac)>=5000 )




#colect garbage
gc()

```


# 5 - rarefaction normalization
```{r}



#check plot for some minimum library sizes 
lapply(unnormalized_BacFun_ps_l, function (x)
  ggplot(data = sample_data(x),
         mapping = aes(x=amendment, y=library_size, color = sample_type, fill = sample_type))+
    geom_jitter()+
    geom_boxplot()+
    geom_hline(yintercept=50000, color="green")+
    geom_hline(yintercept=25000, color="yellow")+
    geom_hline(yintercept=10000, color="red")+
theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1)))



#finally, a rarefaction curve
lapply(unnormalized_BacFun_ps_l, function (x)
  rarecurve(t(as.data.frame(otu_table(x))),
          label = FALSE,
          step = 2000,
          main="Rarefaction", ylab = "Number of ASVs", xlab = "Number of DNA reads",
          abline(v = min(sample_sums(x)), col="red", lwd=3, lty=2)))


```


## 5.1 - rarefy bacteria
```{r}
#rarefy data
min(sample_sums(unnormalized_BacFun_ps_l$Bac)) #minimumum library size
sort.default(colSums(otu_table(unnormalized_BacFun_ps_l$Bac))) # sometimes you might want to lose/remove one or more samples that have a very low library size. you would have to balance the number of samples with the minimum number of sequences. in this particular data set, we won't cut any samples as the rarefaction curve looks very good
set.seed(100) # set a random seed so that whenever you re-run this code you draw the same set of OTUs
rarefied_BacFun_ps_l<-lapply(unnormalized_BacFun_ps_l, function (x)
  rarefy_even_depth(x,
  sample.size = 32750,
  rngseed = FALSE,
  replace = TRUE,
  trimOTUs = TRUE,
  verbose = TRUE))



#check completeness of sample representation
lapply(rarefied_BacFun_ps_l, function (x)
table(sample_data(x)$amendment,
      sample_data(x)$sample_type,
      sample_data(x)$frass_source,
      sample_data(x)$frass_treatment,
      sample_data(x)$frass_dose ))



```

## 5.2 - rarefy fungi
```{r}

min(sample_sums(unnormalized_BacFun_ps_l$Fun)) #minimumum library size
sort.default(colSums(otu_table(unnormalized_BacFun_ps_l$Fun))) # sometimes you might want to lose/remove one or more samples that have a very low library size. you would have to balance the number of samples with the minimum number of sequences. in this particular data set, we won't cut any samples as the rarefaction curve looks very good
set.seed(100) # set a random seed so that whenever you re-run this code you draw the same set of OTUs
rarefied_BacFun_ps_l$Fun<-rarefy_even_depth(unnormalized_BacFun_ps_l$Fun,
                            sample.size = 3997,
                            rngseed = FALSE,
                            replace = TRUE,
                            trimOTUs = TRUE,
                            verbose = TRUE)



#check completeness of sample representation
lapply(rarefied_BacFun_ps_l, function (x)
table(sample_data(x)$amendment,
      sample_data(x)$sample_type,
      sample_data(x)$frass_source,
      sample_data(x)$frass_treatment,
      sample_data(x)$frass_dose ))


```

# 6 - normalize with cummulative sum scaling (CSS)
```{r}

# first, let's transform the phyloseq object into an MR experiment object

MRexp_objt_l<-lapply(unnormalized_BacFun_ps_l, function(x)
  phyloseq_to_metagenomeSeq(x))


#here you can acess the abundance matrix normalized by cummulative sum scaling. you could overwirte the phyloseq object with this
CSS_matrix_l <- lapply(MRexp_objt_l, function(x)
    MRcounts(x, norm = TRUE, log = TRUE)) # using a log scale will essentially reduce the impact of common species and increase the impact of rare species

#make a new phyloseq object list...
CSS_BacFun_ps_l<-unnormalized_BacFun_ps_l

# and now change it's taxa table
CSS_BacFun_ps_l<-mapply (function (x,y){
  otu_table(x)<-otu_table(y, taxa_are_rows = TRUE)
  return(x)
  },
  x = CSS_BacFun_ps_l,
  y = CSS_matrix_l,
  SIMPLIFY = FALSE)

#this is your final phyloseq object
CSS_BacFun_ps_l




```



# 7 - saves the different analysis-ready objects

```{r}

#let's save this  objects as RData so they can be loaded in other scripts
save(unnormalized_BacFun_ps_l, file = "./Data/unnormalized_BacFun_ps_l.RData")


#let's save this rarefied objects as RData so they can be loaded in other scripts
save(rarefied_BacFun_ps_l, file = "./Data/rarefied_BacFun_ps_l.RData")


#let's save these unnormalized objects as RData so they can be loaded in other scripts
save(CSS_BacFun_ps_l, file = "./Data/CSS_BacFun_ps_l.RData")


```

