---
title: "2_beta_diversity"
author: "shaphan"
date: "11/17/2022"
output: html_document
editor_options: 
  
  
  chunk_output_type: console
---


# 0 - Load libraries & data


```{r}


#load necessary libraries
library(phyloseq)
library(ggplot2)
library(vegan)
library(EcolUtils)
library(microbiome)
library(car)
library(agricolae)
library(metagMisc)
library(ggpubr)
library(purrr)
library(EcolUtils)
library(ggrepel)


# increases memory limit used by R. 
# NOTE: This works on R 4.1.2 but not on R 4.2.2. if memory limits become a problem in R 4.2.2, figure a way to increase maximum memory usage
memory.limit(size = 350000)



#load the phyloseq objects containing all microbiome data
load(file = "./Data/CSS_BacFun_ps_l.RData")


```



## 0.1 - separate bulk soil from rhizosphere and frass
```{r}
#split by sample type
CSS_bacfun_ps_sampletype_l<-lapply(CSS_BacFun_ps_l, function (x)
  metagMisc::phyloseq_sep_variable(x, variable = "sample_type"))

# remove bulk soil bac outlier
CSS_bacfun_ps_sampletype_l$Bac$bulk_soil<-subset_samples(CSS_bacfun_ps_sampletype_l$Bac$bulk_soil, sample_names(CSS_bacfun_ps_sampletype_l$Bac$bulk_soil) !="93B")
 
# remove rhizosphere  fun outlier              
CSS_bacfun_ps_sampletype_l$Fun$rhizosphere<-subset_samples(CSS_bacfun_ps_sampletype_l$Fun$rhizosphere, sample_names(CSS_bacfun_ps_sampletype_l$Fun$rhizosphere) !="176R")         
            



```

# 1 - Check beta diversity in a NMDS ordination

Some clustering and treatment effects are evident!
```{r}

# calculate ordinations
NMDS_BacFun_l <-lapply(CSS_bacfun_ps_sampletype_l, function (x)
  lapply(x, function(y)
          ordinate(y,
                   method="NMDS",
                   distance="bray",
                   try=2000,
                   autotransform=TRUE)))



# define a function to calculate and plot NMDS based on bray-curtis
NMDS_calculation_l <- function(physeq_l, ordination_output) {
  
  # this function takes as input a list of phyloseq objects and a list of calculated ordinations
  # it also takes a list of bray-curtis distances calculated with phyloseq::ordinate
  # it returns a list of NMDS plots, with titles in each plot
  
  untitled_plot_l<-mapply(function(x, y)
    phyloseq::plot_ordination(physeq = x, 
                              ordination = y,
                              color ="frass_source",
                              shape = "frass_treatment_and_dose")+
      scale_shape_manual(values=c(1,2,4,16, 17))+
      geom_point(aes(fill = frass_source))+
      theme_classic() +
      labs(subtitle = paste("Stress", round(y$stress, digits=4))) +
      theme(plot.title = element_text(size = 10, face = "bold")) + # extarcts on where this is Bac or Fun, add to plot title
      theme(legend.position="right"),
    x = physeq_l,
    y = ordination_output,
    SIMPLIFY = FALSE)
  
  
  
  # adds firt taxonomic rank of first ps object as part of the tittle (Bac or Fun)
  titles_l <- paste(names(physeq_l), unique(as.character(physeq_l[[1]]@tax_table[,1])))
  
  
  
  Plot_l<- mapply(function (z,w)
    z + ggtitle(w),
    z = untitled_plot_l,
    w = titles_l,
    SIMPLIFY = FALSE)
  
  return(Plot_l)
}

#use custom function, creating a list of plotsfunction
NMDS_Bac_l<-NMDS_calculation_l(physeq_l = CSS_bacfun_ps_sampletype_l$Bac, 
                               ordination_output = NMDS_BacFun_l$Bac )

NMDS_Fun_l<-NMDS_calculation_l(physeq_l = CSS_bacfun_ps_sampletype_l$Fun, 
                               ordination_output = NMDS_BacFun_l$Fun )


# arrange all 3 plots in a pannel

NMDS_Bac_l<-ggarrange(NMDS_Bac_l$bulk_soil,
                        NMDS_Bac_l$rhizosphere,
                        NMDS_Bac_l$frass, common.legend = TRUE,
                          legend = "top")

NMDS_Fun_l<-ggarrange(NMDS_Fun_l$bulk_soil,
                        NMDS_Fun_l$rhizosphere,
                        NMDS_Fun_l$frass, common.legend = TRUE,
                          legend = "top")

#export plot
ggsave(filename = "./Results/NMDS_Bac_3_sample_types.pdf" ,
       plot = NMDS_Bac_l,
       width = 195,
       height = 195,
       units = "mm" )

ggsave(filename = "./Results/NMDS_Fun_3_sample_types.pdf" ,
       plot = NMDS_Fun_l,
       width = 195,
       height = 195,
       units = "mm" )



```


## 1.1 filter samples and recalcuate ordintions

```{r}
# remove frass sample type from list
CSS_bacfun_BulkRhizo_ps_l<-lapply(CSS_bacfun_ps_sampletype_l, function (x) x[c(1,3)])

# remove frass source N from samples
CSS_bacfun_BulkRhizo_ps_l<-lapply(CSS_bacfun_BulkRhizo_ps_l, function(y)
  lapply(y, function(x){
  x<-subset_samples(x, frass_source != "N") # from frass source N
  x<-prune_taxa(taxa_sums(otu_table(x)) > 0, x) # remove ASVs with zero counts (that is, ASVS that only show up on frass source N)
return(x)
}))


# calculate ordinations
NMDS_BacFun_BUlkRhizo_l <-lapply(CSS_bacfun_BulkRhizo_ps_l, function (x)
  lapply(x, function(y)
          ordinate(y,
                   method="NMDS",
                   distance="bray",
                   try=2000,
                   autotransform=TRUE)))



#use custom function, creating a list of plotsfunction
NMDS_Bac_l<-NMDS_calculation_l(physeq_l = CSS_bacfun_BulkRhizo_ps_l$Bac, 
                               ordination_output = NMDS_BacFun_BUlkRhizo_l$Bac )

NMDS_Fun_l<-NMDS_calculation_l(physeq_l = CSS_bacfun_BulkRhizo_ps_l$Fun, 
                               ordination_output = NMDS_BacFun_BUlkRhizo_l$Fun )


# arrange all 3 plots in a pannel
NMDS_BacFun_l<-ggarrange( plotlist = c(NMDS_Bac_l, NMDS_Fun_l),
                          common.legend = TRUE,
                          legend = "top")



#export plot
ggsave(filename = "./Results/NMDS_BacFun_BulkRhizo.pdf" ,
       plot = NMDS_BacFun_l,
       width = 190,
       height =190,
       units = "mm" )






```




# 2 - Ccheck beta dispersion

beta dispersion is like multivatite homogeniety of variances. just like you can't compare properly very small error bars with very large error bars, you should check if the spread of the data clouds is similar per treatment

Beta dispersion of fungal rhizosphere differed by frass source
Beta dispersion of bacterial bulk soil and rhizosphere differed by frass source
other beta dispersions where similar, specially when all 4 factors are falttened and laid next to each other


```{r}

#define function to calculate and plot beta dispersion
beta_disp_plotAndTest<-function(phyloseq_list, group){
  # phyloseq_list = a list of phyloseq objects
  # group = the variable you want to test the beta dispersion of, in quotes
    beta_disper_list<-lapply(phyloseq_list, function (x) 
  betadisper(phyloseq::distance(t(otu_table(x)), method="bray"), sample_data(x)[[group]])) #selects only column "group""

  # gets the names of the list
tiles_list<-names(beta_disper_list)
  
  #runs anova on beta dispersions
get_p<-lapply(beta_disper_list, function (x) 
  anova(x, permutations = 999))


p_dispersion<-map(get_p, 5) # gets the p value of the dispersion test
p_dispersion <- p_dispersion[!is.na(p_dispersion)] # removes the generated NA


#runs anova on beta dispersions
bet_disp_PCOa_plot<-mapply(function (x,y,z) 
  plot(x, 
       main = y, 
       sub =z, 
       xlab="p value for homogeniety test:", 
       ylab="PCoA"),
  x = beta_disper_list,
  y = tiles_list,
  z = p_dispersion,
  SIMPLIFY = FALSE) 


return(bet_disp_PCOa_plot)
}


# execute custom fucntion above

lapply(CSS_bacfun_BulkRhizo_ps_l, function(x) beta_disp_plotAndTest(x, "frass_source"))
lapply(CSS_bacfun_BulkRhizo_ps_l, function(x) beta_disp_plotAndTest(x, "frass_dose"))
lapply(CSS_bacfun_BulkRhizo_ps_l, function(x) beta_disp_plotAndTest(x, "frass_treatment"))
lapply(CSS_bacfun_BulkRhizo_ps_l, function(x) beta_disp_plotAndTest(x, "frass_treatment_and_dose"))




```

# 3 - run PERMANOVA 

We see significant treatment effects, with clear pairwise distinctions, all across the board (bac, fun, rhizo, bulk, interactions...) all experimental factors and their interactions influence microbial community composition in both niches for both kingdoms!

as there are only 2 levels per factor, post-hoc tests are not necessary



```{r}
#testing permanovas 16S and ITS
set.seed(303848)
permanova_BacFun_results <- lapply(CSS_bacfun_BulkRhizo_ps_l, function (y)
  lapply(y, function(x)
       adonis2(formula = phyloseq::distance(t(otu_table(x)), method="bray") ~ frass_treatment*frass_source*frass_dose,
              data = as(sample_data(x),"data.frame"), # changing with as.data.frame is insufficient
              permutations = how(nperm=999))))



```

# 3.1 - run PERMANOVA post-hoc

```{r}

# for bacteria
lapply(CSS_bacfun_BulkRhizo_ps_l$Bac, function (x)
  adonis.pair(dist.mat= phyloseq::distance(otu_table(x), method="bray"),
              Factor= as.factor(as(phyloseq::sample_data(x),"data.frame")$frass_treatment_dose_source)))

# for bacteria
lapply(CSS_bacfun_BulkRhizo_ps_l$Fun, function (x)
  adonis.pair(dist.mat= phyloseq::distance(otu_table(x), method="bray"),
              Factor= as.factor(as(phyloseq::sample_data(x),"data.frame")$frass_treatment)))
```



# 4 - Alpha diversity calculations & plot



```{r}

#load data, lirbary
library("microbiome")
library("car")

load(file = "./Data/rarefied_BacFun_ps_l.RData")

#Calculate richness for root and soil
calculated_diversity_l<-lapply(rarefied_BacFun_ps_l, function (x){
  df<-estimate_richness(x, measures = c("Observed", "Chao1", "Shannon", "Simpson"))
  rownames(df)<-gsub("X", "", rownames(df))
  return(df)
})




#### add diversity metrics to mapping file of phyloseq objects. we do this so we can perform anovas, acess metadat, make nicer plots, etc
# turn diversity into ps metadata
merg_to_ps_l<-lapply(calculated_diversity_l, function (x) sample_data(x))

# merge phyloseq objects, extract them as df
calculated_diversity_df_l<-mapply(function (x,y)
  as(sample_data(merge_phyloseq(x,y)),"data.frame"),
  x =rarefied_BacFun_ps_l,
  y = merg_to_ps_l,
  SIMPLIFY = FALSE )


#Lets see this on a simple boxplot for Shannon diversity index
alpha_plot_nofrass<-lapply(calculated_diversity_df_l, function (x)
  ggplot(dplyr::filter(x, sample_type !="frass"), 
         aes( x =frass_treatment_and_dose, y = Shannon, fill =frass_source, color = frass_source, alpha = 0.5  ))+
  geom_violin(aes(alpha = 0.2))+
  geom_dotplot(binaxis = "y", 
               stackdir = "center",
               position = "dodge",
               binwidth = 0.04, 
               aes( x =frass_treatment_and_dose))+
  theme_bw()+
  scale_y_continuous(expand = c(0, 0), limits = c(0, NA))+
  facet_wrap(~ sample_type )+
  labs(title = paste("Shannon diversity indexes")))

# add kingdom as subtitles
alpha_plot_nofrass$Bac<-alpha_plot_nofrass$Bac + labs(subtitle = "Bacteria")
alpha_plot_nofrass$Fun<-alpha_plot_nofrass$Fun + labs(subtitle = "Fungi")


#save as a plot
ggsave(filename = "./Results/alpha_div_2_sample_types.pdf",
       plot = ggarrange(alpha_plot_nofrass$Bac,
                        alpha_plot_nofrass$Fun),
       width = 30,
       height = 15,
       units = "cm")








#Lets see this on a simple boxplot for Shannon diversity index
alpha_plot_frass<-lapply(calculated_diversity_df_l, function (x)
  ggplot(dplyr::filter(x, sample_type =="frass"), 
         aes( x =frass_treatment_and_dose, y = Shannon, fill =frass_source, color = frass_source, alpha = 0.5  ))+
  geom_violin(aes(alpha = 0.2))+
  geom_dotplot(binaxis = "y", 
               stackdir = "center",
               position = "dodge",
               binwidth = 0.04, 
               aes( x =frass_treatment_and_dose))+
  theme_bw()+
  scale_y_continuous(expand = c(0, 0), limits = c(0, NA))+
  facet_wrap(~ sample_type )+
  labs(title = paste("Shannon diversity indexes")))

# add kingdom as subtitles
alpha_plot_frass$Bac<-alpha_plot_frass$Bac + labs(subtitle = "Bacteria")
alpha_plot_frass$Fun<-alpha_plot_frass$Fun + labs(subtitle = "Fungi")


#save as a plot
ggsave(filename = "./Results/alpha_div_frass_sample_types.pdf",
       plot = ggarrange(alpha_plot_frass$Bac,
                        alpha_plot_frass$Fun),
       width = 25,
       height = 20,
       units = "cm")



```

## 4.1 - remove N and Frass conditions 
```{r}

# remove conditions that make an anova model complicated
calculated_diversity_filtred_df_l<-lapply(calculated_diversity_df_l, function (x)
 dplyr::filter(x, frass_source !="N" & sample_type !="frass"))


#Lets see this on a simple boxplot for Shannon diversity index
alpha_plot<-lapply(calculated_diversity_filtred_df_l, function (x)
  ggplot(x, aes( x =frass_treatment_and_dose, y = Shannon, fill =frass_source ))+
  geom_violin(aes(alpha = 0.3))+
  geom_dotplot(binaxis = "y", stackdir = "center", position = "dodge", binwidth = 0.05, aes( x =frass_treatment_and_dose))+
  theme_bw()+
  facet_wrap(~ sample_type )+
  labs(title = paste("Shannon diversity indexes")))

# add kingdom as subtitles
alpha_plot$Bac<-alpha_plot$Bac + labs(subtitle = "Bacteria")
alpha_plot$Fun<-alpha_plot$Fun + labs(subtitle = "Fungi")

#save as a plot
ggsave(filename = "./Results/alpha_div_BacFun_BulkRhizo.pdf",
       plot = ggarrange(alpha_plot$Bac,
                        alpha_plot$Fun),
       width = 40,
       height = 15,
       units = "cm")

```


# 5 - alpha diversity tests




## 5.1 separate bulk soil and rhizosphere

we have to separate rhizosphere from bulk soil as their homogeneity of varainces are quite different
```{r}

# remove conditions that make an anova model complicated, but keep N control
calculated_diversity_filtred_df_l<-lapply(calculated_diversity_df_l, function (x)
 dplyr::filter(x, sample_type !="frass"))


# for anova models, it's best to separate sample types as they have different variaances
calculated_diversity_filtred_df_l_l<-lapply(calculated_diversity_filtred_df_l, function(x)
  base::split(x, f = x$sample_type))

# 3 sample type factors were inherited form the previous df... this will remove them from the list
calculated_diversity_filtred_df_l_l$Bac$frass<-NULL
calculated_diversity_filtred_df_l_l$Fun$frass<-NULL

```

## 5.2 - check homogeniety of varainces

from "Simpson","Observed","Shannon","Chao1", only untransformed Shannon is homogebenous in both sample types of both Bac and Fun ( 3999 reads in fungi )

```{r}

# check homogeniety of variances aross the nested list, in 4 diversity metrics and two transformations
homogeniety_check<-lapply(calculated_diversity_filtred_df_l_l, function(x) # for both bacteria an fungi...
  lapply(x, function (y) # in both bulk and rhizo..
    apply(y[,c("Simpson","Observed","Shannon","Chao1")],2,function (z){ # run the following function for these varaibles
      untransformed<-leveneTest(z ~ frass_source*frass_treatment*frass_dose, data = y)
      log10_trans<-leveneTest(log10(z) ~ frass_source*frass_treatment*frass_dose, data = y)
      sqrt_trans<-leveneTest(sqrt(z) ~ frass_source*frass_treatment*frass_dose, data = y)
      return(list("untransformed" = untransformed, # returns all transformations in a named list
                  "log10_trans" = log10_trans,
                  "sqrt_trans" = sqrt_trans))
    }
      ) # closes apply
    ) #closes second lapply
  ) # closes first lapply
```

## 5.3 -  test alpha diversity per treatment



```{r}
# anova test
anova_shannon<-lapply(calculated_diversity_filtred_df_l_l, function(x) # for both bacteria an fungi...
  lapply(x, function (y) 
summary(aov(Shannon ~ frass_source*frass_treatment*frass_dose, data = y))))

# poshoc option 1
lapply(calculated_diversity_filtred_df_l_l, function(x) # for both bacteria an fungi...
  lapply(x, function (y) 
TukeyHSD(aov(Shannon ~ frass_source*frass_treatment*frass_dose, data = y))))

# poshoc option 2
tukey_shannon<-lapply(calculated_diversity_filtred_df_l_l, function(x) # for both bacteria an fungi...
  lapply(x, function (y) {
tx <- with(y, interaction(frass_source, frass_treatment, frass_dose)) #needed for tukey to test interaction
aovTukey<-aov(Shannon ~ tx, data = y)#needed for tukey to test interaction
HSD_output<-HSD.test(aovTukey, "tx", group=TRUE, console=TRUE)#post-hoc
return(HSD_output)
}))


```

